scripts/tulipsport_sync.py

```diff
@@ -0,0 +1,275 @@
+        "start_date_local": start_date_local,
+        "average_heartrate": average_heartrate,
+        "map": map,
+        "start_latlng": start_latlng,
+        "distance": distance,
+        "moving_time": moving_time,
+        "elapsed_time": elapsed_time,
+        "average_speed": average_speed,
+        "location_country": location_country,
+    }
+    return namedtuple("activity_db_instance", activity_db_instance.keys())(
+        *activity_db_instance.values()
+    )
+
+
+def find_last_tulipsport_start_time(track_ids):
+    start_time = None
+    tulipsport_ids = [
+        id for id in track_ids if str(id).startswith(TULIPSPORT_FAKE_ID_PREFIX)
+    ]
+    if tulipsport_ids:
+        tulipsport_ids.sort()
+        # 从模拟的构造ID（特殊前缀(666) + 活动开始时间的timestamp + 活动距离(单位：米，支持单次活动最大距离为999,999米)）中读取时间信息
+        start_time = datetime.fromtimestamp(
+            int(str(tulipsport_ids[-1])[3:-6]), DEFAULT_TIMEZONE
+        )
+    return start_time
+
+
+def get_new_activities(token, old_tracks_ids, with_gpx=False):
+    s = requests.Session()
+    headers = {"Authorization": token}
+    activity_summary_list = get_all_activity_summaries(
+        s, headers, find_last_tulipsport_start_time(old_tracks_ids)
+    )
+    activity_summary_list = [
+        activity
+        for activity in activity_summary_list
+        if activity["id"] not in old_tracks_ids
+    ]
+    print(f"{len(activity_summary_list)} new activities to generate")
+    tracks = []
+    if with_gpx and not os.path.exists(GPX_FOLDER):
+        os.mkdir(GPX_FOLDER)
+    old_gpx_ids = os.listdir(GPX_FOLDER)
+    old_gpx_ids = [i.split(".")[0] for i in old_gpx_ids if not i.startswith(".")]
+    for activity_summary in activity_summary_list:
+        activity_id = activity_summary["aid"]
+        print(f"parsing activity id {activity_id}")
+        try:
+            activity_detail = get_activity_detail(s, headers, activity_id)
+            track = merge_summary_and_detail_to_nametuple(
+                activity_summary, activity_detail
+            )
+            tracks.append(track)
+            if with_gpx and activity_summary["id"] not in old_gpx_ids:
+                save_activity_gpx(activity_summary, activity_detail, track)
+        except Exception as e:
+            print(f"Something wrong paring tulipsport id {activity_id} " + str(e))
+    return tracks
+
+
+def save_activity_gpx(summary, detail, track):
+    if not summary["outdoor"]:
+        return
+    gpx = gpxpy.gpx.GPX()
+    gpx.nsmap["gpxtpx"] = "http://www.garmin.com/xmlschemas/TrackPointExtension/v1"
+    gpx_track = gpxpy.gpx.GPXTrack()
+    gpx_track.name = track.name
+    gpx.tracks.append(gpx_track)
+
+    # 详情接口具体的内容参考文档链接：https://open.tulipsport.com/document
+    # map_data_list的结构为[ [latitude, longitude, elevation, section, distance, hr, time, cadence], ... ]
+    point_list = detail["map_data_list"]
+    last_section = ""
+    avg_hr = int(detail["avg_hr"])
+    avg_cadence = int(detail["avg_cadence"])
```

Comment:
I got an error: `Something wrong paring tulipsport id  xxx 'avg_cadence'`
This field may not exist. Could you fix this? @rdshoep

Reply:
OK, I try to add a compatible check. My data works fine~

Reply:
I add the compatible check, please help me to check your data.🤝

If the error still occurs, you can tell me your activityId. I find somebody to check the data.

Reply:
use `detail.__contains__("avg_cadence")` instead of `detail["avg_cadence"]` for the if statement or it would fail right away.

Reply:
Oh, I see. It's more strict than javascript~ Let me try again😅

Reply:
Done～ I test it by `del detail["avg_cadence"]` manually